lock-api/src/main/java/com/palantir/lock/watch/ClientLockWatchEventLogImpl.java

```diff
@@ -0,0 +1,220 @@
+        IdentifiedVersion currentVersion = getLatestVersionAndVerify(toVersion);
+
+        if (!versionInclusive.isPresent() || differentLeaderOrTooFarBehind(currentVersion, versionInclusive.get())) {
+            return TransactionsLockWatchEvents.failure(snapshotUpdater.getSnapshot(currentVersion));
+        }
+
+        if (eventMap.isEmpty()) {
+            return TransactionsLockWatchEvents.success(ImmutableList.of(), timestampToVersion);
+        }
+
+        IdentifiedVersion fromVersion = versionInclusive.get();
+
+        return TransactionsLockWatchEvents.success(
+                new ArrayList<>(eventMap.subMap(fromVersion.version(), true, toVersion.version(), true).values()),
+                timestampToVersion);
+    }
+
+    @Override
+    public synchronized Optional<Set<LockDescriptor>> getEventsBetweenVersions(
+            IdentifiedVersion startVersion,
+            IdentifiedVersion endVersion) {
+        IdentifiedVersion currentVersion = getLatestVersionAndVerify(endVersion);
+        IdentifiedVersion fromVersion = createInclusiveVersion(startVersion);
+
+        if (differentLeaderOrTooFarBehind(currentVersion, fromVersion)) {
+            return Optional.empty();
+        }
+
+        if (eventMap.isEmpty()) {
+            return Optional.of(ImmutableSet.of());
+        }
+
+        List<LockWatchEvent> events =
+                new ArrayList<>(eventMap.subMap(fromVersion.version(), true, endVersion.version(), true).values());
+        Set<LockDescriptor> locksTakenOut = new HashSet<>();
+        events.forEach(event -> locksTakenOut.addAll(event.accept(LockEventVisitor.INSTANCE)));
+
+        return Optional.of(locksTakenOut);
+    }
+
+    @Override
+    public synchronized Optional<IdentifiedVersion> getLatestKnownVersion() {
+        return latestVersion;
+    }
+
+    private boolean differentLeaderOrTooFarBehind(IdentifiedVersion currentVersion, IdentifiedVersion startVersion) {
+        return !startVersion.id().equals(currentVersion.id()) || eventMap.floorKey(startVersion.version()) == null;
+    }
+
+    private IdentifiedVersion createInclusiveVersion(IdentifiedVersion startVersion) {
+        return IdentifiedVersion.of(startVersion.id(), startVersion.version() + 1);
+    }
+
+    private IdentifiedVersion getLatestVersionAndVerify(IdentifiedVersion endVersion) {
+        Preconditions.checkState(latestVersion.isPresent(), "Cannot get events when log does not know its version");
+        IdentifiedVersion currentVersion = latestVersion.get();
+        Preconditions.checkArgument(IdentifiedVersion.comparator().compare(endVersion, currentVersion) > -1,
+                "Transactions' view of the world is more up-to-date than the log");
+        return currentVersion;
+    }
+
+    private void processSuccess(LockWatchStateUpdate.Success success) {
+        Preconditions.checkState(latestVersion.isPresent(), "Must have a known version to process successful updates");
```

Comment:
so there's an interesting class of bug with this sort of thing. Basically, you can partially succeed - accept some but not all of the writes, and then fail (e.g. what if resetWithSnapshot fails below for some reason). Then you're in a corrupt state.

It's relatively straightforward to cover this. Basically, you have a boolean field 'failed' and before you mutate any state, you set it to true, and then false before you're done. Before your reading methods, you checkState(!failed) and if it fails for any reason you know you have a bug.

A different thing is to have immutable state which you replace on success (e.g. make transactional), but that changes your code style a fair bit.